### PR TITLE
feat: Tree refactor with List pieces

### PR DESCRIPTION
**Pull request by ElliotTheHuman**
- Refactored `TreeItem` to utilize `ListItem` underneath.
- Also took the opportunity to refactor `TreeGroup` to make indentation selectors in `TreeStyle` more streamlined

### Requirements

Please check the following items are addressed in your pull request (or are not applicable)

- [ ] Includes test coverage for all changes
- [ ] Documentation updated
- [ ] i18n impacts
- [ ] a11y impacts
- [ ] :rotating_light: Check for image-snapshot changes (run `yarn image-snapshots` locally)

### CODE REVIEW

File: packages/components/src/List/ListItemContext.tsx [hunk 1]
```diff
@@ -25,21 +25,16 @@
...
 
 import { createContext } from 'react'
-import { ListItemDimensions } from './types'
-import { density0 } from './utils'
+import { DensityRamp } from './types'
 
 export interface ListItemContextProps {
-  itemDimensions: ListItemDimensions
+  density: DensityRamp
   iconGutter: boolean
```
Comment (ElliotTheHuman) [inline packages/components/src/List/ListItemContext.tsx:32]: Because `ListItem` now accepts `density` directly, I thought it'd make sense to have the `ListItemContext` provide the density value instead of the `ItemDimensions` object (which was previously looked up in a parent `List` but now is looked up in each item).

Code readability-wise, I think this is cleaner (since the `ListItem` code reads as "If you gave me a density, use that density. Otherwise, use the context density"). However, we might not like the performance hit of each item doing its own dimension lookup.

Thoughts? Happy to retool `ListItemContext` to hold an itemDimensions object if that's the better route.

File: packages/components/src/Tree/Tree.tsx [hunk 2]
```diff
@@ -25,68 +25,117 @@
...
+        ref={detailRef}
+      >
+        {content}
+      </ListItemInnerDetail>
+    ),
+    options: {
+      accessory,
+      hoverDisclosure,
+    },
+  }
```
Comment (ElliotTheHuman) [inline packages/components/src/Tree/Tree.tsx:118]: All this dark magic is used to prevent the Accordion from opening when clicking a `Tree`'s detail element and `accessory === true`.

There's an argument to have this behavior baked into `List`, but I figure that this is specifically a `Tree` problem and so any dark magic belongs within `Tree`'s implementation.

File: packages/components/src/Tree/stories/Tree.story.tsx [hunk 3]
```diff
@@ -30,10 +30,10 @@ import { Tree, TreeProps, TreeItem } from '..'
 
 export * from './BorderRadius.story'
 export * from './ColorfulTree.story'
+export * from './Density.story'
 export * from './DisabledAndSelected.story'
 export * from './FieldPicker.story'
 export * from './FileTree.story'
-export * from './Flat.story'
```
Comment (ElliotTheHuman) [inline packages/components/src/Tree/stories/Tree.story.tsx:36]: Will create a `List` story with a header since that seems more appropriate than "Flat" `Tree`.

File: packages/components/src/Tree/TreeHeading.tsx [hunk 4]
```diff
@@ -25,63 +25,39 @@
...
-  ${TreeItemLabel}, ${TreeGroupLabel}, ${AccordionDisclosure} {
-    ${color}
-  }
+}) => {
+  return (
+    <Paragraph
+      className={className}
+      fontSize="xxsmall"
+      fontWeight="semiBold"
+      lineHeight="xsmall"
```
Comment (ElliotTheHuman) [inline packages/components/src/Tree/TreeHeading.tsx:50]: While it's possible to derive `fontSize` and `lineHeight` from the `Tree`'s density, the Explore team had specific desires about the sizing of the `Tree` group headers.

Keeping this set in stone for now made sense, or at least until someone requests for these to be scalable with density / customizable.

File: packages/components/src/Tree/TreeStyle.tsx [hunk 5]
```diff
@@ -49,46 +52,51 @@ const dividersCSS = css`
...
-        border && generateTreeBorder(depth, theme)}
+      ${({ border, depth, indicatorSize, theme }) =>
+        border && generateTreeBorder(depth, indicatorSize, theme)}
     }
 
     & > ${AccordionDisclosureStyle} {
+      ${ListItem} {
+        font-weight: ${({ branchFontWeight, theme: { fontWeights } }) =>
+          branchFontWeight ? fontWeights.normal : fontWeights.semiBold};
+      }
```
Comment (ElliotTheHuman) [inline packages/components/src/Tree/TreeStyle.tsx:70]: `font-weight` was being clobbered by the browser's default button CSS, so making this selector more specific does the trick (Note: this may not be the ideal solution, happy to discuss better alternatives).

File: packages/components/src/Tree/TreeStyle.tsx [hunk 6]
```diff
@@ -49,46 +52,51 @@ const dividersCSS = css`
...
-    border-width: 0;
-    height: 100%;
-
-    & > ${TreeItemLabel} {
-      background-color: transparent;
-      padding-left: ${({ theme }) => theme.space.none};
-    }
+  & > ${Accordion} > ${AccordionContent} > ${List} > ${TreeHeading},
+  & > ${Accordion} > ${AccordionContent} > ${List} > ${ListItem} > button,
+  & > ${Accordion} > ${AccordionContent} > ${List} > ${ListItem} > a {
```
Comment (ElliotTheHuman) [inline packages/components/src/Tree/TreeStyle.tsx:89]: These selectors (and the ones below) boil down to "Indent my direct child `TreeHeadings` and `ListItem`s

File: packages/components/src/Tree/TreeHeading.tsx [hunk 7]
```diff
@@ -25,63 +25,36 @@
...
-      {truncate ? <Truncate>{label}</Truncate> : label}
-    </TreeGroupLabel>
-    {children}
-  </div>
-)
+}) => {
+  return (
+    <Paragraph className={className} truncate={truncate}>
+      {children}
+    </Paragraph>
```
Comment (lukelooker) [inline packages/components/src/Tree/TreeHeading.tsx:48]: Does this need to be a `li` underneath to keep DOM happy? (immediate descendants of `<ul>` / `<ol>` is usually a `li`)

Comment (lukelooker) [reply packages/components/src/Tree/TreeHeading.tsx:48]: Ok - poked around, appears to not be a problem.

File: packages/components/src/Tree/TreeItem.tsx [hunk 8]
```diff
@@ -24,266 +24,51 @@
...
+      keyColor={keyColor}
+      onBlur={onBlur}
+      onClick={onClick}
+      onKeyDown={onKeyDown}
+      onKeyUp={onKeyUp}
+      selected={selected}
+      {...props}
+    >
+      {children}
+    </ListItem>
```
Comment (lukelooker) [inline packages/components/src/Tree/TreeItem.tsx:70]: This file refactor is awesome 😃

File: packages/components/src/Tree/TreeStyle.tsx [hunk 9]
```diff
@@ -58,37 +77,44 @@ export const TreeStyle = styled.div<TreeStyleProps>`
...
```
Comment (lukelooker) [inline packages/components/src/Tree/TreeStyle.tsx:107]: Nesting could help make this easier to read:


```suggestion
  & > ${Accordion} > ${AccordionContent} > ${List} {
    & > ${TreeHeading}, & > ${ListItem} > button, ${ListItem} > a {
```

File: packages/components/src/Tree/stories/BorderRadius.story.tsx [hunk 10]
```diff
@@ -39,7 +44,8 @@ export const BorderRadiusOverride = () => (
 )
 
 const BorderRadiusOverrideTree = styled(Tree)`
-  ${AccordionDisclosureStyle}, ${TreeItemLabel} {
+  ${AccordionDisclosureStyle},
+  ${Accordion} > ${AccordionContent} > ${List} > ${ListItem} > button {
```
Comment (lukelooker) [inline packages/components/src/Tree/stories/BorderRadius.story.tsx:48]: Would be nice to find a way to make the old CSS work so upstream changes aren't needed / provide a stable styling "contract"

File: packages/components/src/Tree/types.ts [hunk 11]
```diff
@@ -24,53 +24,31 @@
...
-  detailAccessory?: boolean
-  /**
-   * Icon element that appears between the Tree indicator and the Tree label
-   */
-  icon?: IconNames
-  /**
-   * Text label of the Tree
-   * Note: This is a required prop
-   */
-  label: ReactNode
```
Comment (lukelooker) [inline packages/components/src/Tree/types.ts:64]: woohoo!

File: packages/components/src/Tree/types.ts [hunk 12]
```diff
@@ -24,53 +24,31 @@
...
    */
   branchFontWeight?: boolean
   /**
-   * Prevent text wrapping on long labels and instead render truncated text
-   **/
-  truncate?: boolean
+   * Determines the sizing and spacing of child TreeItems
+   * @default 0
+   */
+  density?: DensityRamp
```
Comment (lukelooker) [inline packages/components/src/Tree/types.ts:51]: I wonder if we should just add this to `ListItem` since I could see cases where folks want / need to specify explicitly instead of at the `List` / `Tree` level?

**Review by lukelooker — COMMENTED** [review]
Some minor questions and comments but overall this looks really great!

File: packages/components/src/Tree/TreeStyle.tsx (hunk 9 above)
Comment (lukelooker) [reply packages/components/src/Tree/TreeStyle.tsx:107]: Honestly I'm not sure my refactor here really helps - seems more complex if anything. AKA: Ignore me.

File: packages/components/src/Tree/TreeHeading.tsx (hunk 7 above)
Comment (ElliotTheHuman) [reply packages/components/src/Tree/TreeHeading.tsx:48]: Good call. Technically speaking, this should probably be an `li` based on MDN:
Screen Shot 2021-02-04 at 11 35 17 AM

I read on an SO post that browsers can handle it, but it's pretty simple to get this wrapped in an `li` (Note: it does slightly change the selectors in `TreeStyle` but only by a smidge).

Comment (ElliotTheHuman) [reply packages/components/src/Tree/TreeHeading.tsx]: Resolved in d581d9b

File: packages/components/src/List/ListItem.tsx [hunk 13]
```diff
@@ -90,7 +88,9 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
   const {
     children,
     className,
+    color,
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/List/ListItem.tsx:91]: In this case, the default is just for `icon` (i.e. if given a color, use that color, otherwise, use "text1" for the icon). It does make sense though to not have the color logic sit in the prop where it's hard to parse out. I tried to make things a little more readable (as well as make the color prop work on `Tree`s in [this commit](d581d9b)).

Let me know what you think / how I can make this better!

File: packages/components/src/List/ListItem.tsx [hunk 14]
```diff
@@ -133,7 +143,7 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
     icon || iconArtwork ? (
       <Icon
         artwork={iconArtwork}
-        color="text1"
+        color={color || 'text1'}
```
[the remark this answers is not in the document]
Comment (ElliotTheHuman) [reply packages/components/src/List/ListItem.tsx]: See comment above for additional thoughts

File: packages/components/src/Tree/types.ts (hunk 12 above)
Comment (ElliotTheHuman) [reply packages/components/src/Tree/types.ts:51]: We did actually! We needed a way to pass the density value to the parent-most `Tree`'s inner item, and adding the prop onto `ListItem` made the most sense (but it's a @private so hopefully shouldn't be out in the wild too much).

File: packages/components/src/Tree/stories/BorderRadius.story.tsx (hunk 10 above)
Comment (ElliotTheHuman) [reply packages/components/src/Tree/stories/BorderRadius.story.tsx:48]: That would be ideal... would making it a prop be a viable option? Then it could just be on us to manage the border-radius'ing.

Alternatively, we _could_ try exporting the entire selector 😝  Haven't played around with it much, but in that world, users would just have to worry about importing the "direct child item selector + disclosure selector" and not have to maintain it in anyway.

**Comment by ElliotTheHuman** [comment]
Note: did some reworking of how colors are distributed from Tree and TreeItem / ListItem.

Some of the logic gets hairy, but it's (mostly) all managed via props and not within the styled selectors which seems a little easier to follow. Default color "text5" of `Tree` structures is still set by a selector though.

File: packages/components/src/Tree/TreeHeading.tsx [hunk 15]
```diff
@@ -26,11 +26,15 @@
...
 import { Paragraph } from '../Text'
 
-export interface TreeHeadingProps extends TextColorProps {
+export interface TreeHeadingProps {
+  children: string
   className?: string
+  /**
+   * Determines the text color
+   */
+  color?: string
```
Comment (lukelooker) [inline packages/components/src/Tree/TreeHeading.tsx:37]: Probably ideal to use `TextColorProps`

File: packages/components/src/Tree/stories/BorderRadius.story.tsx (hunk 10 above)
Comment (lukelooker) [reply packages/components/src/Tree/stories/BorderRadius.story.tsx:48]: `ListItemLabel`

File: packages/components/src/List/ListItem.tsx [hunk 16]
```diff
@@ -86,11 +84,26 @@ export interface ListItemProps
...
+  itemRole?: 'link' | 'button'
+}
+
+export const ListItemLabel = styled.div
+  .withConfig<ListItemLabelProps>({
+    shouldForwardProp: (prop) => prop !== 'itemRole',
+  })
+  .attrs<ListItemLabelProps>(({ disabled, itemRole }) => ({
+    as: !disabled && itemRole === 'link' ? 'a' : 'button',
+  }))<ListItemLabelProps>``
```
Comment (ElliotTheHuman) [inline packages/components/src/List/ListItem.tsx:98]: This _should_ keep `itemRole` from getting down to the actual `div` element, correct? I'm not seeing any errors in the console, but wanted to make sure I was using "shouldForwardProp" properly here.

File: packages/components/src/Tree/stories/BorderRadius.story.tsx (hunk 10 above)
Comment (ElliotTheHuman) [reply packages/components/src/Tree/stories/BorderRadius.story.tsx]: Resolved in 8e727fb

File: packages/components/src/Tree/TreeHeading.tsx (hunk 15 above)
Comment (ElliotTheHuman) [reply packages/components/src/Tree/TreeHeading.tsx]: As discussed, we cut `TreeHeading` down to its bare bones (i.e. so that it's only role is to get proper indentation from parent `Tree` and contain children). Now known as `TreeBranch`.

Resolved in a8838ce

File: packages/components/src/Truncate/Truncate.tsx [hunk 17]
```diff
@@ -25,19 +25,26 @@
...
   children: ReactNode
+  className?: string
 }
 
 /**
  * Prevent text wrapping on long labels and instead render truncated text.
  * Renders a tooltip to view the entire text content on hover.
  **/
-export const Truncate: FC<TruncateProps> = ({ children, width = '100%' }) => {
+const TruncateLayout: FC<TruncateProps> = ({
```
Comment (ElliotTheHuman) [inline packages/components/src/Truncate/Truncate.tsx:43]: I needed to be able to change line height on the `Truncate` container in `ListItem` since default button browser styling was clobbering / increasing `ListItem` height by 1px. 

However, couldn't change it directly on `Truncate`'s CSS because that would change the layout on existing `Truncate` use cases (i.e. `FieldCheckboxGroup`).

Decided that exporting `Truncate` as a styled component made sense, but let me know if that's something we totally do not want and I can rethink how we get the proper line-height for `ListItem` here.

File: packages/components/src/List/utils/listItemPadding.ts [hunk 18]
```diff
@@ -42,8 +41,8 @@ export const listItemPadding = ({
    * Without it, density = -1 ListItems would be at 44px.
    */
   const pt = propsPy === '0.375rem' ? propsPy : space[propsPy]
-  const pr = accessory ? '0' : space[propsPx]
   const pb = pt
+  const pr = space[propsPx]
```
Comment (ElliotTheHuman) [inline packages/components/src/List/utils/listItemPadding.ts:45]: We want that inner button / a to always have right padding (afaik). If the button content / label stretches to the edge in the `accessory` case, it would make sense for it to have some right padding rather than reach the edge of the button.

**Comment by ElliotTheHuman** [comment]
Changes since last time:
- `Truncate` is now exported as a styled component for styling purposes within `ListItem`
- Added `ListItemLabel` styled component for easy border radius targeting
- Fixed some minor CSS issues (i.e. border on `TreeArtifical` being weird, `Truncate` + `accessory` not playing nicely because flex)
- `listItemPadding` always has right padding since  properly aligns unhovered, hoverDisclosure, accessory items

File: packages/components/src/List/ListItem.tsx (hunk 16 above)
Comment (lukelooker) [reply packages/components/src/List/ListItem.tsx:111]: Yep - that should swat it away :)

File: packages/components/src/List/utils/listItemPadding.ts (hunk 18 above)
Comment (ElliotTheHuman) [reply packages/components/src/List/utils/listItemPadding.ts:45]: Previous implementation:
`const pr = accessory ? '0' : space[propsPx]`

File: packages/components/src/List/ListItemLayoutAccessory.tsx [hunk 19]
```diff
@@ -26,31 +26,31 @@
...
-import { FlexItem } from '../Layout'
+import { Flex } from '../Layout'
 import { ListItemLayoutProps } from './ListItemLayout'
 import { listItemPadding } from './utils'
 
 export const ListItemLayoutAccessoryInternal: FC<ListItemLayoutProps> = ({
   children,
   className,
-  containerCreator,
+  labelCreator,
```
Comment (lukelooker) [inline packages/components/src/List/ListItemLayoutAccessory.tsx:36]: What's the reason to do this as a function instead of as a component?

File: packages/components/src/List/ListItemLayoutAccessory.tsx [hunk 20]
```diff
@@ -26,31 +26,31 @@
...
         {description}
-      </FlexItem>
+      </Flex>
     </>
   )
 
   return (
     <>
-      {containerCreator({
+      {labelCreator({
```
Comment (lukelooker) [inline packages/components/src/List/ListItemLayoutAccessory.tsx:53]: Same as above - but why not do this as `<LabelCreator className={className}>content</LabelCreator>`?

File: packages/components/src/Truncate/Truncate.tsx [hunk 21]
```diff
@@ -58,10 +65,17 @@ export const Truncate: FC<TruncateProps> = ({ children, width = '100%' }) => {
     width: 'auto',
   })
 
+  const className = join([domProps.className, propsClassName], ' ')
```
Comment (lukelooker) [inline packages/components/src/Truncate/Truncate.tsx:68]: I'd just use built-in join here. 

```suggestion
  const { className, otherDomProps } = domProps
  const className = [className, propsClassName].join(' ')
```

File: packages/components/src/Truncate/Truncate.tsx [hunk 22]
```diff
@@ -58,10 +65,17 @@ export const Truncate: FC<TruncateProps> = ({ children, width = '100%' }) => {
...
+
   return (
     <>
       {tooltip}
-      <TextStyle {...domProps} ref={textRef} width={width}>
+      <TextStyle
+        {...omit(domProps, 'className')}
```
Comment (lukelooker) [inline packages/components/src/Truncate/Truncate.tsx:74]:
```suggestion
        {...otherDomProps}
```

File: packages/components/src/Truncate/Truncate.tsx [hunk 23]
```diff
@@ -25,19 +25,26 @@
  */
 import React, { FC, useState, useCallback, ReactNode } from 'react'
 import styled from 'styled-components'
+import join from 'lodash/join'
+import omit from 'lodash/omit'
```
Comment (lukelooker) [inline packages/components/src/Truncate/Truncate.tsx:29]: Down with `lodash` (see below)

File: packages/components/src/Truncate/Truncate.tsx (hunk 17 above)
Comment (lukelooker) [reply packages/components/src/Truncate/Truncate.tsx:43]: I'm cool with exporting this but we should log a ticket to eventually fix `Truncate` to have `line-height: 1` (fixing the locations that depend on that) because it really does seem more "correct" that way and then we could unwind `Tree`'s need to reach-in as it were.

**Review by lukelooker — APPROVED** [review]

File: packages/components/src/List/ListItemLayoutAccessory.tsx [hunk 24]
```diff
@@ -26,31 +26,31 @@
...
     <>
       {icon}
-      <FlexItem>
+      <Flex flexDirection="column" minWidth={0}>
         {children}
         {description}
-      </FlexItem>
+      </Flex>
     </>
   )
```
Comment (lukelooker) [inline packages/components/src/List/ListItemLayoutAccessory.tsx:50]: const LabelCreator = labelCreator

File: packages/components/src/List/ListItemLayoutAccessory.tsx (hunk 20 above)
Comment (ElliotTheHuman) [reply packages/components/src/List/ListItemLayoutAccessory.tsx:53]: Discussed over meet, but because `LabelCreator` needs to wrap around the content in the `XLayout` components, we would need to pass in the component via a prop, then set it to a variable with a capital name so that we can create JSX elements off of it.

IIRC, the plan is to keep things as they are for now, but might be worth revisiting this assuming there's no performance loss from passing in a component vs. a vanilla function.

**Comment by ElliotTheHuman** [comment]
Excuse the latest commit name 🤦  I set the density to -3 by default, the commit was just named incorrectly by yours truly.

File: packages/components/src/List/ListItem.tsx [hunk 25]
```diff
@@ -86,11 +97,26 @@ export interface ListItemProps
...
+
 const ListItemInternal: FC<ListItemProps> = (props) => {
   const {
     children,
     className,
+    color,
     current,
+    density: propsDensity,
```
Comment (lukelooker) [inline packages/components/src/List/ListItem.tsx:119]:
```suggestion
    density: propsDensity = -3,
```

Comment (lukelooker) [reply packages/components/src/List/ListItem.tsx:119]: Can you switch the default density back down to match the previous setup so we can compare screenshots?

**Review by lukelooker — CHANGES_REQUESTED** [review]
Push density down so we can do one last careful comparison of snapshot changes :)

File: packages/components/src/List/ListItem.tsx (hunk 25 above)
Comment (ElliotTheHuman) [reply packages/components/src/List/ListItem.tsx:119]: Ah I set the default density back to `-3` on `Tree` in particular since those stories were at that small "Developer" density.

`ListItem` and `List` default to `0` on main, but I can bump those down to `0` as well if you'd like! Just wanted to make sure.

Comment (lukelooker) [reply packages/components/src/List/ListItem.tsx:119]: Oh - I'm a dork and misread here. Ignore me!

**Review by lukelooker — APPROVED** [review]
Looking at `File Tree Closed` is appears the icon size and spacing changed a little bit:

image

image

That said the Field Picker appears to be rock-solid so I think we're good to go!
